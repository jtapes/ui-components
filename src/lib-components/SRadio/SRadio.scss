@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-radio {
  display: inline-flex;
  flex-direction: column;
  justify-content: center;
  cursor: pointer;

  &__input {
    position: relative;
    display: flex;
    flex-shrink: 0;
    margin-top: $size--4;
    width: $size--16;
    height: $size--16;
  }

  &__label {
    display: flex;
    margin-left: $size--8;
  }

  &__description {
    margin-left: $size--8;
  }

  &__inner {
    position: absolute;

    &::after {
      content: '';
      position: absolute;
      border-radius: 50%;
      width: $size--8;
      height: $size--8;
      background-color: $color--primary;
      transform: scaleY(0);
      transition: scale $animate--time;
    }
  }

  &__original {
    appearance: none;
    border: 1px solid $color--font-muted;
    border-radius: 50%;
    width: 100%;
    height: 100%;
    transition: $animate--time;
    cursor: pointer;

    &:hover {
      border-color: $color--primary-accent;
    }
  }

  &:not(&--disabled):active &__original {
    border-color: transparent;
    box-shadow: 0 0 0 2px $color--outline;
  }

  &--grid {
    border: 1px solid $color--border;
    padding: $size--20;
    transition: $animate--time;

    &:hover {
      border-color: $color--primary-accent;
    }

    &:active {
      border-color: transparent;
      box-shadow: 0 0 0 2px $color--outline;
    }

    &.s-radio {
      &--checked {
        border-color: transparent;
        box-shadow: 0 0 0 2px $color--outline;

        &.s-radio--disabled {
          .s-radio__inner {
            &::after {
              background-color: $color--border;
            }
          }
        }
      }

      &--disabled {
        border: 1px solid $color--border;
        box-shadow: none;
        background-color: $color--background;
        cursor: not-allowed;

        .s-radio {
          &__original {
            border-color: transparent;
            background-color: $color--border;
          }
        }
      }

      &--contented {
        background-color: $color--white;
      }
    }
  }

  &--checked {
    .s-radio {
      &__label {
        color: $color--primary;
      }

      &__original {
        border-color: $color--primary;
      }

      &__inner::after {
        transform: scaleY(1) translate(50%, 50%);
      }
    }

    &.s-radio--disabled {
      .s-radio__original {
        border-color: $color--border;
        background-color: transparent;
      }
    }
  }

  &--disabled {
    cursor: not-allowed;

    .s-radio {
      &__original {
        border-color: transparent;
        background-color: $color--border;
        cursor: not-allowed;
      }

      &__inner {
        &::after {
          background-color: $color--border;
        }
      }

      &__label {
        color: $color--font-muted;
      }
    }
  }
}
