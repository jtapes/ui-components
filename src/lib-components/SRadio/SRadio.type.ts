export type SRadioValue = object | number | string

export enum SRadioStates {
  DISABLED = 'disabled',
  ERROR = 'error',
  DEFAULT = 'default'
}
