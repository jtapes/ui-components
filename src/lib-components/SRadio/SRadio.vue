<template>
  <label
    class="s-radio text-sm"
    :class="[
      {
        's-radio--checked': value === checkedValue,
        's-radio--contented': $slots.content,
      },
      `s-radio--${mode}`,
      `s-radio--${state}`,
    ]"
  >
    <div class="d-f">
      <div class="s-radio__input">
        <div class="s-radio__inner" />
        <input
          class="s-radio__original"
          type="radio"
          aria-hidden="false"
          :value="checkedValue"
          :checked="value === checkedValue"
          :disabled="state === 'disabled'"
          @change="handleModelChange"
        />
      </div>
      <div class="s-radio__label" :class="{ bold: mode === 'grid' }">
        <!--  @slot Слот для контента checkbox   -->
        <slot>{{ label }}</slot>
      </div>
      <div
        v-if="$slots.description"
        class="s-radio__description"
        :class="{ 'color-font-muted': state === 'disabled' }"
      >
        <!-- @slot Слот для доп. описания к label -->
        <slot name="description" />
      </div>
    </div>
    <div v-if="mode === 'grid' && $slots.content" class="s-radio__content">
      <!-- @slot Слот для контента в radio с модом grid -->
      <slot name="content" />
    </div>
  </label>
</template>

<script lang="ts">
import { Component, Vue, Model, Prop } from 'vue-property-decorator'

import type { SRadioValue } from '@/lib-components/SRadio/SRadio.type'

/** Компонент для выбора одного из значений */
@Component
export default class SRadio extends Vue {
  @Model('change')
  value!: SRadioValue

  /** Значение radio button */
  @Prop()
  checkedValue!: SRadioValue

  /** Ярлык checkbox */
  @Prop(String)
  label?: string

  /** Тип радио кнопки */
  @Prop({ type: String, default: 'default' })
  mode!: string

  /** Состояние радио ("disabled", "error") */
  @Prop({ type: String, default: 'default' })
  state!: string

  handleModelChange(): void {
    this.$emit('change', this.checkedValue)
  }
}
</script>
