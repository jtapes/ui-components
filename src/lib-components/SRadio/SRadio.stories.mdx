import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs'
import SRadio from '@/lib-components/SRadio/SRadio.vue'
import results from '/.jest-test-results.json';

<Meta
  title="DESIGN SYSTEM/Atom/Radio"
  component={SRadio}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    label: { control: 'text' },
    state: {
      control: { type: 'select' },
      options: [
        'disabled',
        'error',
        'default',
      ]
    },
    mode: {
      options: [
        'default',
        'grid'
      ],
      control: { type: 'select' },
    }
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-Market-UI-Kit?node-id=846%3A1100',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Radio button

<Description of={SRadio} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SRadio },
  template:
    `
    <div>
      <div>
        <s-radio v-model="selectedValue" checked-value="0" :mode="mode" :label="label" :state="state" @change="handleChange" />
      </div>
      <div>
        <s-radio v-model="selectedValue" checked-value="1" :mode="mode" :label="label" :state="state" @change="handleChange" />
      </div>
    </div>
`,
	data() {
		return {
			selectedValue: null
		}
	},
	methods: {
		handleChange() {
			action('change')()
		}
	}
})

export const GridTemplate = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SRadio },
  template:
    `
    <div class="d-f">
      <div>
        <s-radio v-model="selectedValue" checked-value="0" :mode="mode" :label="label" :state="state" @change="handleChange" />
      </div>
      <div class="ml-40">
        <s-radio v-model="selectedValue" checked-value="1" :mode="mode" :label="label" :state="state" @change="handleChange" />
      </div>
    </div>
`,
  data() {
    return {
      selectedValue: null
    }
  },
  methods: {
    handleChange() {
      action('change')()
    }
  }
})

export const templateCustomLabel = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SRadio },
  template:
    `
    <div class="d-f">
      <s-radio v-model="selectedValue" checked-value="0" :mode="mode" :label="label" :state="state" @change="handleChange">
        <div class="w-100 mt-4-n">
          <p class="text-sb bold">Р/с: 40702810123450101230</p>
          <p class="text-sm color-font-secondary">
          Московский банк ПАО Сбербанк г. Москва <br/>
          БИК: 044521234<br/>
          К/с: 30101234500000000225
          </p>
        </div>
      </s-radio>
      <div class="ml-40">
        <s-radio v-model="selectedValue" checked-value="1" :mode="mode" :label="label" :state="state" @change="handleChange">
        <div class="w-100 mt-4-n">
          <p class="text-sb bold">Р/с: 40702810123450101228</p>
          <p class="text-sm color-font-secondary">
          Московский банк ПАО Сбербанк г. Москва <br/>
          БИК: 044521234<br/>
          К/с: 30101234500000000225
          </p>
        </div>
      </s-radio>
      </div>
    </div>
`,
  data() {
    return {
      selectedValue: null
    }
  },
  methods: {
    handleChange() {
      action('change')()
    }
  }
})

<Canvas withToolbar withSource="none">
  <Story name="SRadio" args={{ label: 'Radio' }}>
    {Template.bind({})}
  </Story>
  <Story name="SRadio (grid)" args={{ label: 'Radio', mode: 'grid' }}>
    {GridTemplate.bind({})}
  </Story>
  <Story name="SRadio (custom label)" args={{ label: 'Radio', mode: 'grid' }}>
    {templateCustomLabel.bind({})}
  </Story>
</Canvas>

## Пример

<Source
  language="html"
  code={`
    <s-radio v-model="selectedValue" mode="default" label="Radio" :disabled="false" />
  `}
/>

<ArgsTable of={SRadio} />
