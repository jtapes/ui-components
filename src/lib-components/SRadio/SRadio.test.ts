import { mount } from '@vue/test-utils'
import SRadio from '@/lib-components/SRadio/SRadio.vue'

describe('SRadio', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SRadio, {
      propsData: {
        value: 1,
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
  test('Is emitted by $emit()', async () => {
    const wrapper = mount(SRadio, {
      propsData: {
        value: 1,
      },
    })

    const input = wrapper.find('input[type="radio"]')
    await input.setChecked()

    expect(wrapper.emitted('change')).toBeTruthy()
  })
  test('Label of radio exist label prop', () => {
    const label = 'Radio'

    const wrapper = mount(SRadio, {
      propsData: {
        label,
      },
    })

    const labelContent = wrapper.find('.s-radio__label')

    expect(labelContent.text()).toBe(label)
  })
})
