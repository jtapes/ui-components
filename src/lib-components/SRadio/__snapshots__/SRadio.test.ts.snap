// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SRadio Is Vue snapshot 1`] = `
<label
  class="s-radio text-sm s-radio--default s-radio--default"
>
  <div
    class="d-f"
  >
    <div
      class="s-radio__input"
    >
      <div
        class="s-radio__inner"
      />
       
      <input
        aria-hidden="false"
        class="s-radio__original"
        type="radio"
        value=""
      />
    </div>
     
    <div
      class="s-radio__label"
    >
      
    </div>
     
    <!---->
  </div>
   
  <!---->
</label>
`;
